config/README.md

```diff
@@ -166,6 +167,226 @@ fmt.Println(m.World)
+type Calculator struct {
+  Left  int
+  Right int
+  Op    Operation
+}
+
+func (c Calculator) Eval() int {
+  return c.Op(c.Left, c.Right)
+}
+```
+
+Calculator constructor needs only `config.Provider` and it loads configuration from
+the root:
+
+```go
+func NewCalculator(cfg Provider) (*Calculator, error){
+  calc := &Calculator{}
+  return calc, cfg.Get(Root).Populate(calc)
+}
+```
+
+`Operation` has a  function type, but we can make it configurable. In order for
+a provider to know how to deserialize it, `Operation` type needs to implement
+`text.Unmarshaller` interface:
+
+```go
+func (o *Operation) UnmarshalText(text []byte) error {
+  switch s := string(text); s {
+  case "+":
+    *o = func(left, right int) int { return left + right }
+  case "-":
+    *o = func(left, right int) int { return left - right }
+  default:
+    return fmt.Errorf("unknown operation %q", s)
+  }
+
+  return nil
+}
+```
+
+Testing it with a static provider will be easy, we can define all arguments there
+with the expected result:
+
+```go
+func TestCalculator_Eval(t *testing.T) {
+  t.Parallel()
+
+  table := map[string]Provider{
+    "1+2": NewStaticProvider(map[string]string{
+      "Op": "+", "Left": "1", "Right": "2", "Expected": "3"}),
+    "1-2": NewStaticProvider(map[string]string{
+      "Op": "-", "Left": "2", "Right": "1", "Expected": "1"}),
+  }
+
+  for name, cfg := range table {
+    t.Run(name, func(t *testing.T) {
+      calc, err := NewCalculator(cfg)
+      require.NoError(t, err)
+      assert.Equal(t, cfg.Get("Expected").AsInt(), calc.Eval())
+    })
+  }
+}
+```
+
+We should not forget to test the error path as well:
+
+```go
+func TestCalculator_Errors(t *testing.T) {
+  t.Parallel()
+
+  _, err := newCalculator(NewStaticProvider(map[string]string{
+    "Op": "*", "Left": "3", "Right": "5"
+  }))
+
+  require.Error(t, err)
+  assert.Contains(t, err.Error(), `unknown operation "*"`)
+}
+```
+
+For integration/E2E testing you can customize `config.Loader` to load
+configuration files from either custom folders(`Loader.SetDirs()`),
+or custom files(`Loader.SetFiles()`), or register new providers on top of
+existing providers(`Loader.RegisterProviders()`) that will override values
+of default configs.
+
+## Utilities
+
+`Config` package comes with several helpers that can make writing tests,
+create new providers or amend existing ones much easier.
+
+* `NewCachedProvider(p Provider)` returns a new provider that wraps `p`
+  and caches values in underlying map. It also registers callbacks to track
+  changes in all values it cached, so you can call `cached.Get("something")`
+  and don't worry about latencies much. It is safe for concurrent use by
+  multiple goroutines.
+
+* `MockDynamicProvider` is a mock provider that can be used to test dynamic
+  features, it implements `Provider` interface and lets you to set values
+  to trigger change callbacks.
+
+* Sometimes dynamic providers let you to register only one callback per key.
```

Comment:
- Sometimes dynamic providers only let you register one callback per key
- Instead of "sometimes," be explicit, like: "The dynamic provider above only lets you register…"
- Tell us which provider you're talking about and in which scenarios they restrict you to one callback per key